### PR TITLE
skeleton and rpc connections code for fixed curp system

### PR DESCRIPTION
**Pull request by Schwidola0607**
## Draft
Make RPCs for backups, master, and witnesses. 
Currently fixed system i.e no support for failure handler / leader re-election. Nodes role are fixed. No dynamic membership (join/leave) 

Debts: 
- Failure handler / leader re-election can be implemented shortly after this version (not too complex)
- Dynamic membership / reconfiguration is hard. Most likely need to halt entire/majority of system. Task for much later on

Next steps to reach milestone 1:
- Filled in logic for the RPCs 
- Implement simple app that sit on top of Curp

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	filename := args[2]
+	node_map := Parse(filename)
+	master_map, witness_map, _ := ParseByRole(node_map)
+	master := Map_begin(master_map)
+	master_client, _ := rpc.Dial("tcp", master.Ip+":"+master.Port)
+	var witness_clients []*rpc.Client
+	for _, v := range witness_map {
+		c, _ := rpc.Dial("tcp", v.Ip+":"+v.Port)
+		witness_clients = append(witness_clients, c)
+	}
```
Comment (AydanPirani) [inline client/client.go]: Super nitpicky, but we can probably multithread this/do it concurrently

File: curp/master.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * RPC types
+ */
+type ExecuteArgs struct {
+}
+type ExecuteReply struct {
+}
+type SyncArgs struct {
+}
+type SyncReply struct {
+}
```
Comment (AydanPirani) [inline curp/master.go]: Can we rename these structs? RN they sound like functions, which lwk confused me for a bit
```suggestion
type ExecuteArgsData struct {
}
type ExecuteReplyData struct {
}
type SyncArgsData struct {
}
type SyncReplyData struct {
}
```

File: curp/master.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+func (m *Master) Sync(args SyncArgs, reply *SyncReply) error { // syncRPC called by clients
+	return nil
+}
+
+/**
+ * Non-RPC functions
+ */
+func (m *Master) AsyncOrdering() {
```
Comment (AydanPirani) [inline curp/master.go]: imo rename this to a verb-based function 
```suggestion
func (m *Master) OrderAsync() {
```

File: curp/node.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	connection_map := make(map[string]*rpc.Client)
+	for _, v := range node_map {
+		for {
+			addr := v.Ip + ":" + v.Port
+			conn, err := rpc.Dial("tcp", addr)
+			if err != nil {
+				// log.Println("Failed to connect to " + v.name)
+				time.Sleep(150 * time.Millisecond)
+			} else {
+
```
Comment (AydanPirani) [inline curp/node.go:23]: nit but delete
```suggestion
```

File: curp/master.go (hunk 2 above)
Comment (Schwidola0607) [reply curp/master.go]: These are go RPC shenanigans, every RPC call need to pass in a "condensed" <FuncName>args, <FuncName>reply

File: shared/connections.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	"bufio"
+	"log"
+	"os"
+	"strings"
+)
+
+type NodeRole string
+
+const (
+	RoleMaster  NodeRole = "Master"
```
Comment (AydanPirani) [inline shared/connections.go]: All caps for enum?

Comment (Schwidola0607) [reply shared/connections.go]: Go syntax is upper case = public, lower case = private (it's stupid I know)

Comment (AydanPirani) [reply shared/connections.go]: or well, RoleMaster -> ROLE_MASTER

Comment (Schwidola0607) [reply shared/connections.go]: oh sure

File: shared/operation.go [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package shared
+
+// probably just a serialized string
+type Operation struct {
+}
```
Comment (AydanPirani) [inline shared/operation.go]: what does this represent? just an operation enum + data?

File: client/client.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	if len(os.Args) != 3 {
+		fmt.Println("[usage]: " + args[0] + " <identifier> <configuration file>")
+		os.Exit(1)
+	}
+
+	name := args[1]
+	filename := args[2]
+	node_map := Parse(filename)
+	master_map, witness_map, _ := ParseByRole(node_map)
+	master := Map_begin(master_map)
```
Comment (AydanPirani) [inline client/client.go]: why do we have a map if it's just one thing? can we not just use a singular entry for this and pass by reference?

File: shared/operation.go (hunk 6 above)
Comment (Schwidola0607) [reply shared/operation.go]: Element in a log of operations. Basically what has been executed/synced/unsynced as defined by Curp. Left room for potential more struct members but a serialized string probably works fine.

File: client/client.go (hunk 7 above)
Comment (Schwidola0607) [reply client/client.go]: Yeah it's lazy code-reusing, will change.

File: client/client.go (hunk 1 above)
Comment (Schwidola0607) [reply client/client.go]: `rpc.Dial` call blocks and need mutex lock for append so multi-threading won't speed it up. Actual multi-thread happens some line further down.

**Pull request merged**